Broadcom pi zero2w neopixel misbehaving/crash fix

RetiredWizard opened the pull request:
These changes fix adafruit/circuitpython#5947. I've tested on both the zero2w and the pi4b (The 4b didn't exhibit the original issue) and the boards now behave properly with 1 to 30 pixels. I've run a 30 pixel animation on the zero2w for 8+ hours without a hang.


**Original corrupted rgb transmission:**
![image](https://user-images.githubusercontent.com/29083569/218270769-95a59a8e-dfc9-4c9f-b141-66b5920a43a7.png)
**Early Fix, not using transmission queue:**
![image](https://user-images.githubusercontent.com/29083569/218244620-86df6d65-d197-41c1-94b8-11d5d05e78c3.png)
**This PR:**
![image](https://user-images.githubusercontent.com/29083569/218270243-91d54478-9c7e-4c4c-b0ef-1b11fed2b133.png)

RetiredWizard commented:
I just tested this on a zero w rather than the zero2w and it works but it looks like the code is getting stuck more frequently than the other tested board in the loop that checks for an empty queue. I'm not sure that's a problem if I reduce the loop abort counter but I want to test on the three models I have. I'll update once I've finished testing.

RetiredWizard commented:
Tested on 4b, zero2w and zero w. Runs without glitches or crashes.

tannewt commented on ports/broadcom/common-hal/neopixel_write/__init__.c:
```diff
@@ -45,7 +45,10 @@ void common_hal_neopixel_write(const digitalio_digitalinout_obj_t *digitalinout,
     // two.
-    while (port_get_raw_ticks(NULL) < next_start_raw_ticks) {
+    int icnt;
+    while ((port_get_raw_ticks(NULL) < next_start_raw_ticks) &
```
```suggestion
    while ((port_get_raw_ticks(NULL) < next_start_raw_ticks) &&
```

tannewt commented on ports/broadcom/common-hal/neopixel_write/__init__.c:
```diff
@@ -92,7 +95,9 @@ void common_hal_neopixel_write(const digitalio_digitalinout_obj_t *digitalinout,
         COMPLETE_MEMORY_READS;
-        while (CM_PWM->CS_b.BUSY == 0) {
+        icnt = 0;
+        while ((CM_PWM->CS_b.BUSY == 0) & (icnt++ < 1000)) {
```
```suggestion
        while ((CM_PWM->CS_b.BUSY == 0) && (icnt++ < 1000)) {
```

tannewt reviewed (changes requested):
Please use `&&` in your conditions instead of `&`. `&` is a bitwise test and `&&` is boolean logic.

RetiredWizard commented:
So I misunderstood what `COMPLETE_MEMORY_READS` was doing and clearly over used it in the update for the zero w. I've pulled all the calls I added for that update out and I'm testing it now. The only one remaining that I added was at the very end of the code just before the PWM function of the pin is turned off. I feel like that one might be needed but I guess I'll do some tests.
